packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -569,17 +593,40 @@ mod default {
     #[cfg(any(all(rust_1_65, feature = "std"), feature = "spantrace"))]
     use alloc::format;
     use alloc::{vec, vec::Vec};
+    use core::any::TypeId;
     #[cfg(all(rust_1_65, feature = "std"))]
     use std::backtrace::Backtrace;
 
     #[cfg(feature = "spantrace")]
     use tracing_error::SpanTrace;
 
     use crate::{
-        fmt::{hook::HookContext, Emit},
+        fmt::{
+            hook::{into_boxed_hook, BoxedHook, HookContext},
+            Emit,
+        },
         Frame,
     };
 
+    // TODO: this is called for every invocation of call() if we haven't `.insert()` yet
```

Comment:
Can this be solved using [`Once`](https://doc.rust-lang.org/std/sync/struct.Once.html)? We could add the builtin hooks the first time `install_debug_hook` is called, then I think we don't need any special handling 🙂

Reply:
Once you call `install_debug_hook` **once**, it automatically inserts the hooks (after that, this function is never called again), then that's done, but if you `Debug` before that for every `Debug`, we call `builtin()` once. I could potentially solve this via a `RwLock` which acts as a poor man's `OnceCell`.

We need to do this because otherwise, you will only get backtrace and spantrace handling when you call `install_debug_hook` once, which isn't always the case. (Some might just want to leave it as is)

Reply:
Can't we simply add
```rust


    fn install_builtin_hooks() {
        static INSTALL_BUILTIN: Once = Once::new();
        INSTALL_BUILTIN.call_once(|| {
            #[cfg(all(rust_1_65, feature = "std"))]
            Report::install_debug_hook::<Backtrace>(|backtrace, context| {
                let idx = ctx.increment();

                ctx.attach_snippet(format!("Backtrace No. {}\n{}", idx + 1, backtrace));

                vec![Emit::defer(format!(
                    "backtrace with {} frames ({})",
                    backtrace.frames().len(),
                    idx + 1
                ))]
            });

            #[cfg(feature = "spantrace")]
            Report::install_debug_hook::<Backtrace>(|backtrace, context| {
                let idx = ctx.increment();

                let mut span = 0;
                spantrace.with_spans(|_, _| {
                    span += 1;
                    true
                });

                ctx.attach_snippet(format!("Span Trace No. {}\n{}", idx + 1, spantrace));

                vec![Emit::defer(format!(
                    "spantrace with {span} frames ({})",
                    idx + 1
                ))]
            });
        })
    }
```

and call it from `install_debug_hook()` and `get_debug_format_hook()`?

Reply:
100%, will do something similar :D (use existing functions instead because no_std)

Reply:
So instead of `install_debug_hook` you will use `FMT_HOOK.write().insert(...)`?

Reply:
We noticed offline, that `SpanTrace` requires `std`. We decided to use this PR to simplify the hooks module by putting `#[cfg(feature = "std)]` on the whole module.

Reply:
fixed in [e538e4c](https://github.com/hashintel/hash/pull/987/commits/e538e4c8528bfef9be173b0de2a127959e322a83)